add blog improve clang-doc blog posts (#69)
llvm/llvm-blog-www: content/posts/2024-12-04-improve-clang-doc.md

## Work Done

The original scope of the project was to improve the output of clang-doc's generation. However during testing the tool was significantly slower than expected which made developing features for the tool impossible. 
Each compilation of the LLVM codebase was taking upwards of 10 hours on my local machine. Additionally the tool utilized a lot of memory and was prone to crashing with an out of memory error. Similar tools such as Doxygen and Hdoc ran in comparatively less time for the same codebase. This pointed to a significant bottleneck within clang-doc’s codepath when generating large scale software projects. Due to this the project scope quickly changed to improving the runtime of clang-doc so that it could run much faster. It was only during the latter half of the project did the scope change back to improving clang-doc’s generation.

### Added More Test Cases to Clang-Doc test suite

> Clang-doc previously had test’s which did not test the full scope of the tool especially with regards to the HTML or Markdown output of the tool. In order to make sure the optimization experiments were not accidentally breaking the tool, it was necessary to add more end to end tests. 

Reviewer: ```suggestion
Clang-Doc previously had tests which did not test the full scope of the the HTML or Markdown output. I added more end-to-end tests to make sure that in the process of optimizing documentation generation we were not degrading the quality or functionality of the tool. 
```

> In summary I added four comprehensive tests which covered all features that we were not testing such as testing the generation for Enums, Namespace and Records for HTML and Markdown. 

Reviewer: ```suggestion
In summary, I added four comprehensive tests which covered all features that we were not testing such as testing the generation for Enums, Namespace and Records for HTML and Markdown. 
```

### Improve clang-doc’s performance by 1.58 times 

> Internally the way clang-doc works is by leveraging libtooling's ASTVisitor class to parse the declaration in each project and serializing it into an internal format, which gets deserialized later when we output the final format. The bottleneck in clang-doc lied in the way clang-doc was doing redundant work when it was visiting each declaration. 

Reviewer: ```suggestion
Internally, the way Clang-Doc works is by leveraging libtooling's ASTVisitor class to parse the source level declarations in each TU and serializing it into an internal format which gets deserialized later when we output the final format. Eventually, we were able to identify a major bottleneck in clang-doc's performance to doing redundant work when it was processing each declaration. We settled on a caching/memoization strategy to minimize the redundant work.
```

Whatever capitalization you decide on for `clang-doc`, just make it consistent.

 I'd also reword the section a bit more, and include more detail. Clang-doc is parsing each source file/TU into an AST via libtooling, then visiting important nodes in the AST to collect type and source location information for the various declarations. It's also visiting comment nodes in the AST an collecting all of those in to the various *Info bits you worked with. From there the serialized bitcode representation of the documentation Info is collected via a Map-Reduce algorithm to generate the final documentation. This process -- which you should be familiar with -- should be explained here.

For example if we had a the following project: 

```cpp
//File: Base.cpp

class Base {}
```

```cpp
//File: A.cpp

class A : public Base {}
```

```cpp
//File: B.cpp

class B : public Base {}
```

> In this case the ASTVisitor invoke by clang-doc would visit serialized the Base class three times, once when it is parsing Base.cpp, another when its visiting A.cpp then B.cpp. This means any C++ project that heavily leverages inheritance would result in a lot of redundant work. 

Reviewer: ```suggestion
In this case, the ASTVisitor invoked by clang-doc would visit the serialized Base class three times, once when it is parsing Base.cpp, another when its visiting A.cpp then B.cpp. This means any C++ project that heavily leverages inheritance would result in a lot of redundant work. 
```

The problem isn't inheritance, per se. A major part of it is that class/type decls are usually in headers, which appear in many TUs, and none of those are deduplicated or recognized. Visiting nodes w/in a class hierarchy multiple times is unfortunate, but if it was only a problem proportional to the depth of the class hierarchy, we wouldn't have seen such a dramatic performance improvement. The basic problem is that we never checked if we'd seen an item before and did the full processing on every instance. The memoization strategy stopped us from doing redundant things, and we got a big speedup.

The optimization ended up being a simple memoization dictionary which kept track of a list of declaration that clang-doc had visited. 

Here is a plot of the benchmarking numbers:

<div style="margin:0 auto;">
  <img src="/img/clang-doc-benchmark-numbers.png"><br/>
</div>

### Added Template Mustache HTML Backend

> Clang-doc originally used an ad-hoc method of generating HTML, in this project I introduced a templating language as a way of reducing project complexity and reducing the ease of development. Two RFCs were made before arriving at the idea of introducing Mustache as a library. Originally the idea was to introduce a custom templating language, however upon further discussion it was decided that the complexity of designing and implementing a new templating language was too much. 

Reviewer: ```suggestion
Clang-doc originally used an ad-hoc method of generating HTML. I introduced a templating language as a way of reducing project complexity and reducing the ease of development. Two RFCs were made before arriving at the idea of introducing Mustache as a library. Originally the idea was to introduce a custom templating language, however upon further discussion it was decided that the complexity of designing and implementing a new templating language was too much. 
```
Link your RFCs.

> A LLVM community member suggested I implement a mustache as templating language. 

Reviewer: ```suggestion
A LLVM community member suggested using mustache as templating language. 
```

> Mustache was the ideal templating language, since it was very simple to implement, and has a well defined spec that fit what was needed for clang-doc’s use case. The feedback to the RFC was generally positive however there was some a bit of pushback in regards to adding an HTML support library to LLVM. 

Reviewer: ```suggestion
Mustache was the ideal choice since it was very simple to implement, and has a well defined spec that fit what was needed for clang-doc’s use case. The feedback to the RFC was generally positive however there was some a bit of pushback in regards to adding an HTML support library to LLVM. 
```

The bit about pushback is a bit misleading. There was some resistance, but the community member who raised the objection also wasn't aware that we already generate HTML in a number of places, and the new templating language could be used to simplify other cases as well. I think its important to put these things into context if you're going to talk about them. It's also important to chose words that make it clear we respect our colleagues opinions, even when we disagree. I don't think you've done anything to convey otherwise so far, but since I'm asking for this to be rephrased, I want to emphasize that point.

> In terms of engineering wins, this library was able to cut the direct down on HTML backend significantly dropping 500 lines of code. This library was also designed for general purpose use around LLVM, since there are numerous places in LLVM where various tools generate html in its own way. Using the mustache templating library would be a nice way to standardize the codebase. 

Reviewer: What does "cut the direct down on HTML backed" mean? do you mean that we were able to drop 500 lines of C++ from clang-doc? or something else.

### Improve Clang-Doc HTML Output

> The previous version of clang-doc’s output was a pretty minimal bare bones implementation. It had a sidebar that contained every single declaration within the project which created a large unnavigable UI. Typedef documentation was missing, plus method documentation was missing details such as whether or not the method was a const or virtual. There was no linking between other declarations in the project and there was no syntax highlighting on any language construct.

Reviewer: Please check the punctuation and spelling.

With the new mustache changes an additional backend was added using the specifier (--format=mhtml). That addresses these issues. 

Below is a comparison of the same output between the two backends

<div style="margin:0 auto;">
  <img src="/img/clang-doc-old-html-output.png"><br/>
</div>

<div style="margin:0 auto;">
  <img src="/img/clang-doc-new-output.png"><br/>
</div>

You can also visit the output project on my personal github.io page link
[here](https://peterchou1.github.io/).

Note: this output is still a work in progress.

## Learning Insight

> I've learned a lot in the past few months, thanks to GSOC I now have a much better idea of what it’s like to participate in a large open source project. I received a lot of feedback through PR’s, making RFC and collaborating with other GSOC members. I’d learned a lot about how to interact with the community and solicit feedback. I also learned a lot about instrumentation/profiling code having conducted many experiments in order to try to speed clang-doc up.

Reviewer: I think it would have been great to detail those experiments a bit more in an earlier section.

## Future Work

> I plan to work on clang-doc until an MVP product can be generated and evaluated for the LLVM project. My remaining tasks include landing the mustache support library and clang-doc’s mustache backend, as well as gathering feedback from the LLVM community regarding clang-doc’s current output. Additionally, I intend to add test cases for the mustache HTML backend to ensure its robustness and functionality.

Reviewer: You are now one of `clagn-doc`'s official maintainers. You may want to mention that.

---
author: "Peter Chou"
date: "2024-12-04"
tags: ["GSoC", "Clang-Doc"]
title: "GSoC 2024: Improve Clang Doc"
---

Hi, my name is Peter, and this year I was involved in Google Summer of Code 2024. I worked on [improving the Clang-Doc documenation generator](https://discourse.llvm.org/t/improve-Clang-Doc-usability/76996)

Mentors: Petr Hosek and Paul Kirth

## Project Background

Clang-Doc is a documentation generator developed on top of libtooling, as an
alternative to Doxygen. Development started in 2018 and continued through 2019,
however, it has since stalled. Currently, the tool can generate HTML, YAML, and markdown but the generated output has usability issues. This GSOC project aimed to address the pain points regarding the output of the HTML, by adding support for various C++ constructs and reworking the CSS of the HTML output to be more user-friendly.

## Work Done

The original scope of the project was to improve the output of Clang-Doc's generation. However during testing the tool was significantly slower than expected which made developing features for the tool impossible. 

> Each compilation of the LLVM codebase was taking upwards of 10 hours on my local machine. Additionally, the tool utilized a lot of memory and was prone to crashing with an out-of-memory error. Similar tools such as Doxygen and Hdoc ran in comparatively less time for the same codebase. This pointed to a significant bottleneck within Clang-Doc’s code path when generating large-scale software projects. Due to this the project scope quickly changed to improving the runtime of Clang-Doc so that it could run much faster. It was only during the latter half of the project did the scope changed back to improving Clang-Doc’s generation.

Reviewer: ```suggestion
Documentation generation for the LLVM codebase was taking upwards of 10 hours on my local machine. Additionally, the tool utilized a lot of memory and was prone to crashing with an out-of-memory error. Similar tools such as Doxygen and Hdoc ran in comparatively less time for the same codebase. This pointed to a significant bottleneck within Clang-Doc’s code path when generating large-scale software projects. Due to this the project scope quickly changed to improving the runtime of Clang-Doc so that it could run much faster. It was only during the latter half of the project did the scope changed back to improving Clang-Doc’s generation.
```

### Added More Test Cases to Clang-Doc test suite

Clang-Doc previously had tests which did not test the full scope of the the HTML or Markdown output. I added more end-to-end tests to make sure that in the process of optimizing documentation generation we were not degrading the quality or functionality of the tool. 

In summary, I added four comprehensive tests that covered all features that we were not testing such as testing the generation for Enums, Namespace, and Records for HTML and Markdown.

### Improve Clang-Doc’s performance by 1.58 times 

Internally, the way Clang-Doc works is by leveraging libtooling's ASTVisitor class to parse the source level declarations in each TU. 

> Clang-Doc is architected using a Map-Reduce pattern. Clang-Doc parses each fragment of a declaration into an in-memory data format which is serialized then into an internal format and stored as a key value paired, identified by their [USR](https://clang.llvm.org/doxygen/group__CINDEX__CURSOR__XREF.html#ga51679cb755bbd94cc5e9476c685f2df3). After, Clang-Doc deserializes and combines each of the fragment declarations back into the in-memory data format which is used by each of the backend to generate the results.

Reviewer: Nit: every sentence shouldn't start w/ `Clang-Doc`. `It' may be appropriate for some, but you could equally rephrase as `we` or use another literary construct altogether. The point is, it reads as very repetitive.

Many experiments were conducted to identified the source of the bottleneck. First I tried benchmarking the code with many different codebases such JSON, and fmtlib to identify certain code patterns that slowed the code path down. This didn't really work since the bottlenecking only showed up for large codebases like LLVM.
Next I leverage Windows prolifer (since I was coding on windows) however the visualizations was not helpful and the my system was not capable of profiling the 10 hour runtime required to compile LLVM documenation. 

Eventually, we were able to identify a major bottleneck in Clang-Doc's by leveraging the TimeProfiler code to identify where the performance bottleneck was. Clang-Doc was performing redundant work when it was processing each declaration. We settled on a caching/memoization strategy to minimize the redundant work.

For example, if we had the following project: 

```cpp
//File: Base.cpp

class Base {}
```

```cpp
//File: A.cpp

#include "Base.cpp"

...

```

```cpp
//File: B.cpp

#include "Base.cpp"

...
```

> In this case, the ASTVisitor invoked by Clang-Doc would visit the serialized Base class three times, once when it is parsing Base.cpp, another when its visiting A.cpp then B.cpp. The problem was that there was no mechanism to identify declarations that we had already seen. The optimization ended up being a simple memoization dictionary which kept track of a list of declaration that Clang-Doc had visited. 

Reviewer: ```suggestion
In this case, the ASTVisitor invoked by Clang-Doc would visit the serialized Base class three times, once when it is parsing Base.cpp, another when its visiting A.cpp then B.cpp. The problem was that there was no mechanism to identify declarations that we had already seen. Using a simple dictionary which kept track of a list of declaration that Clang-Doc had visited as a basic form of memoization ended up being a surprisingly effective optimization. 
```
Maybe this is better phrasing? the previous one sounded awkward to my ear.

> Here is a plot of the benchmarking numbers:

Reviewer: Perhaps talk about the plot a bit over just saying "here it is"?

<div style="margin:0 auto;">
  <img src="/img/Clang-Doc-benchmark-numbers.png"><br/>
</div>

### Added Template Mustache HTML Backend

Clang-Doc originally used an ad-hoc method of generating HTML. I introduced a templating language as a way of reducing project complexity and reducing the ease of development. Two RFCs were made before arriving at the idea of introducing Mustache as a library. Originally the idea was to introduce a custom templating language, however, upon further discussion, it was decided that the complexity of designing and implementing a new templating language was too much.

> An LLVM community member suggested using Mustache as a templating language. 

Reviewer: I think it would be nice to mention/credit them.

Mustache was the ideal choice since it was very simple to implement, and has a well defined spec that fit what was needed for Clang-Doc’s use case. The feedback on the [RFC](https://discourse.llvm.org/t/rfc-add-template-mustache-language-to-the-support-library/82439/18) was generally positive. While there was some resistance regarding the inclusion of an HTML support library in LLVM, this concern stemmed partly from a lack of awareness that HTML generation already occurs in several parts of LLVM. Additionally, the introduction of Mustache has the potential to simplify other HTML-related use cases.
In terms of engineering wins, this library was able to cut the direct down on the HTML backend significantly dropping 500 lines of code compared to the original Clang-Doc HTML backend. This library was also designed for general-purpose use around LLVM since there are numerous places in LLVM where various tools generate HTML in its way. Using the Mustache templating library would be a nice way to standardize the codebase. 

### Improve Clang-Doc HTML Output

> The previous version of Clang-Doc’s output was a pretty minimal bare bones implementation. It had a sidebar that contained every single declaration within the project which created a large unnavigable UI. Typedef documentation was missing, plus method documentation was missing details such as whether or not the method was a const or virtual. There was no linking between other declarations in the project and there was no syntax highlighting on any language construct.

Reviewer: ```suggestion
The previous version of Clang-Doc’s HTML output was a pretty minimal, bare bones implementation. It had a sidebar that contained every single declaration within the project which created a large unnavigable UI. Typedef documentation was missing, plus method documentation was missing details such as whether or not the method was a const or virtual. There was no linking between other declarations in the project and there was no syntax highlighting on any language construct.
```

Up to you on this, but I'd suggest discussing the UI more since you feel strongly about it. For instance you mention that the UI was hard to use, but I feel like you can lean in here and say more about the what/why. This is a personal style choice though, so it's your call.

---
author: "Peter Chou"

> date: "2024-12-04"

Reviewer 2: ```suggestion
date: "2024-12-23"
```

Many experiments were conducted to identified the source of the bottleneck. First I tried benchmarking the code with many different codebases such JSON, and fmtlib to identify certain code patterns that slowed the code path down. This didn't really work since the bottlenecking only showed up for large codebases like LLVM.
Next I leverage Windows prolifer (since I was coding on windows) however the visualizations was not helpful and the my system was not capable of profiling the 10 hour runtime required to compile LLVM documenation. 

Eventually, we were able to identify a major bottleneck in Clang-Doc's by leveraging the TimeProfiler (similar to -ftime-trace in clang) code to identify where the performance bottleneck was. Clang-Doc was performing redundant work when it was processing each declaration. We settled on a caching/memoization strategy to minimize the redundant work.

For example, if we had the following project: 

```cpp
//File: Base.cpp

class Base {}
```

```cpp
//File: A.cpp

> #include "Base.cpp"

Reviewer 2: Normally one does not include .cpp file. So, the problem was mostly around headers?

...

```

```cpp
//File: B.cpp

#include "Base.h"

...
```

> In this case, the ASTVisitor invoked by Clang-Doc would visit the serialized Base class three times, once when it is parsing Base.cpp, another when its visiting A.cpp then B.cpp. The problem was that there was no mechanism to identify declarations that we had already seen. Using a simple dictionary which kept track of a list of declaration that Clang-Doc had visited as a basic form of memoization ended up being a surprisingly effective optimization. 

Reviewer 2: ```suggestion
In this case, the ASTVisitor invoked by Clang-Doc would visit the serialized Base class three times, once when it is parsing Base.h, another when its visiting A.cpp then B.cpp. The problem was that there was no mechanism to identify declarations that we had already seen. Using a simple dictionary which kept track of a list of declaration that Clang-Doc had visited as a basic form of memoization ended up being a surprisingly effective optimization. 
```
